### ISSUE TITLE
-Tag,  -ExcludeTag, and  -TestName parameters are executing all .Tests.ps1 files

### ISSUE THREAD

**Issue by Rmpanga** [pester/Pester#770]
I have 10 Tests.ps1 scripts with each script has a Describe block. In one of the Tests.ps1 scripts (lets call its A.Tests.ps1)  I specified a tag called UnitTest

 `Describe "A Test"  -Tags @("UnitTest")`

When I run

`Invoke-Pester  -Tag @("UnitTest")`

It runs all 10 Tests.ps1 scripts. I expected it to run only A.Tests.ps1 Describe block which has the matching tag

Note: I have tried using "UnitTest" as well but still the same result.

I'm I doing something wrong or is this an issue ?

Furthermore: I am using Pester 3.4.0

**Comment by Rmpanga** [comment]
After some investiagation I realized the issue.  When you run  

`Invoke-Pester -Tag @("UnitTest")` 

It will run All the scripts with .Tests.ps1 extensions and display the result of each test in the powershell console, just like Write-Host. So, it seems that the -Tag parameter is not working, since the description states that,

> When calling Invoke-Pester, it is possible to specify a -Tag parameter which will only execute Describe blocks containing the same Tag

This statement is incorrect. 'Invoke-Pester -Tag "Some-tag"'  does execute all the Describe blocks. However the -Tag parameter is used to filter out the executed tests when the -PassThru parameter is enabled. 

In other words, 'Invoke-Pester -Tag "Some-Tag" -PassThru' will executed every .Tests.ps1 file. The object that gets "Passed Thru" will be filtered to only include Describe blocks with "Some-Tag"

This is all true for ExcludeTag and TestName

**Comment by it-praktyk** [comment]
I understand your concerns. 

@dlwyatt, [is it true](https://github.com/pester/Pester/issues/461#issuecomment-174714645) for V4 too?

**Comment by it-praktyk** [comment]
Probably answers why it happens and why it can't be implemented in a different way you can find in threads:
- #721 
- #426

**Comment by nohwnd** [comment]
@Rmpanga Just tried it and your description is not accurate. All the `*.tests.ps1 ` are fact exectuted, but each top-level describe inspects the tags and runs only if the tags match. See the implementation [here](https://github.com/pester/Pester/blob/master/Functions/Describe.ps1#L122). 

Here some example code to highlight how it works:
```powershell
# save in t.tests.ps1
Describe "u1" -Tags @("u1") {
    it "abc" {}
}

Describe "u2" -Tags @("u2") {
    Write-Host "I would print if I would run."
    it "abc" {}
}

"I print because I am outside of Describe"
Describe "u3" -Tags @("u3") {
    it "abc" {}
}

# run in console
Invoke-Pester -Path .\t.tests.ps1 -Tag "u1"
```

```
Executing all tests in .\t.tests.ps1 with Tags u1

Executing script .\t.tests.ps1

  Describing u1
    [?] abc 20ms
I print because I am outside of Describe
Tests completed in 20ms
Tests Passed: 0, Failed: 0, Skipped: 0, Pending: 1, Inconclusive: 0 
```

Closing this issue, because it works as specified.  If you need more assistastance feel free to reopen. 😃
